docs/fendermint/general_cross_messages.md

```diff
@@ -0,0 +1,120 @@
+    IpcMsgKind kind;
+    /// @dev Destination of the message.
+    IPCAddress to;
+    /// @dev Address of the sender.
+    IPCAddress from;
+    /// @dev Outgoing nonce for the envelope, ensuring uniqueness per subnet and preventing replay attacks.
+    uint64 localNonce;
+    /// @dev Value being sent with the message (e.g., tokens).
+    uint256 value;
+    /// @dev ABI-encoded message payload.
+    bytes message;
+    /// @dev Original nonce from the source subnet, useful for tracing messages across subnets.
+    uint64 originalNonce;
+}
+```
+
+### Message Types of Interest
+
+- **Call**: Represents a request sent to a destination contract.
+- **Result**: Represents the response to a `Call` message, which could either be the result or an error report.
+
+### Contract Requirements
+
+Both sending and receiving contracts must implement the `IpcExchange` abstract contract from the IPC SDK to handle cross messages. Key methods to implement include:
+
+#### Handling Incoming Calls
+
+This must be implemented in the contract receiving the message:
+
+```solidity
+function _handleIpcCall(
+    IpcEnvelope memory envelope,
+    CallMsg memory callMsg
+) internal virtual returns (bytes memory);
+```
+
+#### Handling Results
+
+This must be implemented in the contract sending the message to handle results from the destination contract:
+
+```solidity
+function _handleIpcResult(
+    IpcEnvelope storage original,
+    IpcEnvelope memory result,
+    ResultMsg memory resultMsg
+) internal virtual;
+```
+
+For an example, refer to [CrossMessengerCaller.sol](../../contracts/contracts/examples/CrossMessengerCaller.sol).
+
+## Results
+
+Results provide either error propagation or responses from the destination contract back to the caller. This creates a request-response mechanism where a `Call` represents the request and a `Result` represents the response.
+
+The IPC envelope will have a `kind` of `Result`, and the message will follow the standardized schema below (see [CrossNet.sol](../../contracts/contracts/structs/CrossNet.sol)):
+
+```solidity
+struct ResultMsg {
+    /// @dev ID of the envelope the result belongs to.
+    bytes32 id;
+    /// @dev Outcome of the call (success or type of error).
+    OutcomeType outcome;
+    /// @dev ABI-encoded return value or failure reason.
+    bytes ret;
+}
+```
+
+### Outcome Types
+
+- **Ok**: Message was delivered successfully, and the `ret` field contains the response from the called contract.
+- **SystemErr**: An error occurred during message transmission. The return value indicates the `InvalidXnetMessageReason` (see [IPCErrors.sol](../../contracts/contracts/errors/IPCErrors.sol)).
+- **ActorErr**: A custom error returned by the called contract.
+
+## How It Works
+
+1. A custom contract calls `sendContractXnetMessage` on the local `Gateway` with a message inside the IPC envelope (of kind `Call`). The envelope includes the destination subnet ID, which is used for routing.
```

Comment:
Hm, the envelope contains the destination _address_, which is prefixed with the destination _subnet path_.